### ISSUE TITLE
No Recursive Step - Potential Cause for Failing Key Tests

### ISSUE THREAD

**Issue by JRJurman** [tunnckoCore/demo-minmorph#14]
When stepping through the logic for this test 
https://github.com/tunnckoCore/demo-minmorph/blob/d587917be0f6e6ae2430a85082fd55b4899758a3/src/app/test-keyed.js#L148-L159

I noticed that we swap a larger chunk of the dom than we need to.  
Specifically, here
![image](https://user-images.githubusercontent.com/326557/36575123-26276732-1817-11e8-9a7e-d622d2ac6ee6.png)
![image](https://user-images.githubusercontent.com/326557/36575130-37147274-1817-11e8-9b2e-945ed7868b6f.png)

We dump all of `futureStartNode` in the parentNode, and then later we pop off the second node.
In reality, I think we need to recurse down the tree more, and then only swap the TEXT nodes. There doesn't appear to be any recursive step - like in nanomorph, here: 
https://github.com/choojs/nanomorph/blob/ccbffcf1f0138a0293f7c71ee6521c7640df7173/index.js#L121-L132

**Comment by olstenlarck** [comment]
Exactly. I was thinking about that and tried few things locally.

**Comment by olstenlarck** [comment]
Hey @JRJurman, @WebReflection!

Heads up! :tada: Code and tooling is now more organized, cleaned up and linted - `yarn install`, `yarn test` & `yarn bench`.

There are only 5 assertions that fails, which represents to of the 2 keyed tests. And it's probably because that recursion problem. 

@WebReflection, do you have any ideas for support such thing in `domdiff`? 
Currently I copy-pasted it here, because the need for `same()` check instead of just `===` in some places.

I marked the failing tests with `// TODO: PRs welcome` for more easier review.

Any ideas and thoughts about the speed and tests are welcome! :penguin:

**Comment by WebReflection** [comment]
@olstenlarck I've no idea what is this about, or what you are asking me. Are you using domdiff? Did you file a bug in domdiff for whatever problem you have encountered?

**Comment by olstenlarck** [comment]
No, i didn't open issues there, but i will in some point. @JRJurman descibed well what we are talking about and i don't know what to add more to that. I'm not sure where the recursion should happen too.

I just can't do currently a demo only using `domdiff`. As about this repo, it's a refactored `nanomorph` with using `domdiff` for the childrens diffing instead.

The `src/domdiff.js` is exactly the same as the latest `domdiff` 0.4.2, except changes in [this](https://github.com/WebReflection/domdiff/blob/master/index.js#L52), [this](https://github.com/WebReflection/domdiff/blob/master/index.js#L58) and [this](https://github.com/WebReflection/domdiff/blob/master/index.js#L62) check, to  checks more than `==` - e.g. `node.isSameNode`, `node.nodeName` equality, keys and etc - [found here](https://github.com/tunnckoCore/demo-minmorph/blob/master/src/domdiff.js#L143-L153).

**Comment by JRJurman** [comment]
@olstenlarck I can work on those changes on domdiff and we can pull in that fork

**Comment by JRJurman** [comment]
Okay, I'm in the process of including an extra parameter that allows for a custom isSameNode function, which defaults to '=='.

However I noticed this comment by WebReflection
https://github.com/choojs/nanomorph/issues/85#issuecomment-350834055

> domdiff also wouldn't care less about patches and it works linearly with a list of nodes, it's not suitable for nested nodes too, just lists of the same parent node.

I believe this is the root cause of this issue

**Comment by WebReflection** [comment]
I think 3 calls to `same(...)` shouldn't be a huge issue, or a relevant performance impact, but I don't like much the idea of adding an extra parameter to that signature, since there are already two optionals at the end, one of which is already a callback.

I'll think about releasing a major version with breaking signature changes that will probably look like this:

```js
futureNodes = domdiff(
  parentNode,     // where changes happen
  currentNodes,   // Array of current items/nodes
  futureNodes,    // Array of future items/nodes (returned)
  extras = {
    compareNode: (a, b) => a === b,
    getNode: node => node,
    beforeNode: null
  }
);
``` 

would that work for you needs?

**Comment by olstenlarck** [comment]
> I believe this is the root cause of this issue

@JRJurman, exactly. The problem is that i don't know where the recursion can (if can) happen - i'm not so deep into this. It would be enough to allow calling some callback where it detects that there is some more child nodes on some of the items on the list.

> I'll think about releasing a major version with breaking signature changes that will probably look like this:

@WebReflection, sounds good yea. Too much params aren't good thing, so start `extras` is good way.

> would that work for you needs?

I think so, partially. The `compareNode` name sounds good or probably plural - `compareNodes`?

In any way, it's good start :)

**Comment by olstenlarck** [comment]
Sweeet! :tada:  I'll try it out.

But the recursion is still the thing. Any ideas of how can be done? How you handle such cases in `hyperhtml` where some of the list items has child nodes?

**Comment by WebReflection** [comment]
You can check the asNode function inside Update.js shared object. Components are all based on children by their own.

**Comment by WebReflection** [comment]
FYI this is the part you might be interested in:
https://github.com/WebReflection/hyperHTML/blob/master/esm/objects/Updates.js#L35-L48

Basically, the `getNode` receives an integer that specifies the operation that is going to be performed. you can pass the right node to insert or remove, and if you control a list of node, you can pass just the first one or the last one and remove all others when needed/asked.

I hope that helps. This is also explained in the domdiff readme.
